packages/template-retail-react-app/app/hooks/use-datacloud.js

```diff
@@ -0,0 +1,461 @@
+/*
+ * Copyright (c) 2025, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import {useMemo} from 'react'
+import logger from '@salesforce/retail-react-app/app/utils/logger-instance'
+import {initDataCloudSdk} from '@salesforce/cc-datacloud-typescript'
+import {getConfig} from '@salesforce/pwa-kit-runtime/utils/ssr-config'
+import {useUsid, useCustomerId, useCustomerType, useDNT} from '@salesforce/commerce-sdk-react'
+import useMultiSite from '@salesforce/retail-react-app/app/hooks/use-multi-site'
+import {useCurrentCustomer} from '@salesforce/retail-react-app/app/hooks/use-current-customer'
+
+export class DataCloudApi {
+    constructor({site}) {
+        this.site = site
+
+        // Grab Data Cloud configuration values and intialize the sdk
+        const {
+            app: {dataCloudAPI: config}
+        } = getConfig()
+
+        const {appSourceId, tenantId} = config
+        let sdk = null
+
+        if (!appSourceId || !tenantId) {
+            console.error('DataCloud API Configuration is missing.')
+            return
+        } else {
+            sdk = initDataCloudSdk(tenantId, appSourceId)
+        }
+        this.sdk = sdk
+    }
+
+    /**
+     * Constructs the base event object with the necessary data required
+     * for every event sent to Data Cloud.
+     *
+     * @param {object} args - The arguments containing event-specific details
+     * @returns {object} - The base event object
+     */
+    _constructBaseEvent(args) {
+        return {
+            guestId: args.guestId,
+            siteId: this.site,
+            sessionId: args.deviceId, //get dwsid from cookie?
+            deviceId: args.deviceId,
+            dateTime: new Date().toISOString(),
+            ...(args.customerId && {customerId: args.customerId}), // Can remove the conditionality after the hook -> Promise is changed in future PWA release
+            ...(args.customerNo && {customerNo: args.customerNo})
+        }
+    }
+
+    _constructUserDetails(args) {
+        return {
+            isAnonymous: args.isGuest,
+            firstName: args.firstName,
+            lastName: args.lastName
```

Comment:
Does it matter if args.[propName] is undefined and get passed into dataCloud?

Reply:
Yes - it is okay - the field is just left empty in Data Cloud